src/cdomains/apron/linearTwoVarEqualityDomain.apron.ml

```diff
@@ -0,0 +1,883 @@
+          done;
+          !result
+      in
+      let all_are_const_in_eq_class zts start size : bool = 
+        let result = ref true in
+        for i = start to start + size - 1 do
+          if not (is_const zts.(i)) then result := false;
+        done;
+        !result
+      in
+      let assign_vars_in_const_eq_class ats zts start size least_i least_b =     
+        for i = start to start + size - 1 do
+          match zts.(i) with
+          | (ai, t1, t2) -> if Z.equal (diff t1 t2) (Z.zero) then ats.(i) <- (ai, t1)
+            else
+              match t1 with
+              | (_, bj) -> ats.(i) <- (ai, (Some least_i, Z.sub bj least_b))
+        done
+      in
+      let assign_vars_in_non_const_eq_class ats zts start size least_i least_b = 
+        for i = start to start + size - 1 do
+          match zts.(i) with
+          | (ai, t1, _) -> 
+            let bj = const_offset t1 in
+            ats.(i) <- (ai, (Some least_i, Z.sub bj least_b))
+        done
+      in
+      match zts with
+      | None -> None
+      | Some zts' ->
+        let result = Array.make (Array.length zts') (0, (None, Z.zero)) in
+        let i = ref 0 in
+        while !i < Array.length zts' do 
+          let n = size_of_eq_class zts' !i in 
+          (if n = 1 then
+             let ztsi = zts'.(!i) in
+             match ztsi with
+             | (i', t1, t2) -> if is_const ztsi && Z.equal (diff t1 t2) (Z.zero) then 
+                 result.(!i) <- (i', (None, const_offset t1))
+               else result.(!i) <- (i', (Some i', Z.zero))
+           else
+             let (least_i, least_b) = least_index_var_in_eq_class zts' !i n in
+             (if all_are_const_in_eq_class zts' !i n then
+                assign_vars_in_const_eq_class result zts' !i n least_i least_b
+              else assign_vars_in_non_const_eq_class result zts' !i n least_i least_b);
+          ); 
+          i := !i + n;
+        done;
+        Some result
+    in
+    let strip_annotation ats = 
+      match ats with
+      | None -> None
+      | Some ats' -> Some (Array.map snd ats')
+    in
+    let join_d t1 t2 =
+      let zipped = ts_zip t1 t2 in
+      sort_z_by_expr zipped;
+      let annotated = process_eq_classes zipped in
+      sort_annotated annotated;
+      let result = strip_annotation annotated in
+      result
+    in
+    if is_bot_env a then b else if is_bot_env b then a else
+      match Option.get a.d, Option.get b.d with
+      | x, y when is_top a || is_top b -> let new_env = Environment.lce a.env b.env 
+        in (identity new_env)
+      | x, y when (Environment.compare a.env b.env <> 0) ->
+        let sup_env = Environment.lce a.env b.env in
+        let mod_x = dim_add (Environment.dimchange a.env sup_env) x in
+        let mod_y = dim_add (Environment.dimchange b.env sup_env) y in
+        {d = join_d mod_x mod_y; env = sup_env}
+      | x, y when EArray.equal x y -> {d = Some x; env = a.env}
+      | x, y  -> {d = join_d x y; env = a.env}
```

Comment:
This seems to be 140 lines of join algorithm?! I would guess the amount of code can be reduced bay an order of magnitude here. Have you considered the helpers that Array, BatArray, List, BatList, Option, and BatOption offer?

If you find that operations on arrays are cumbersome in this case, and you can avoid random-accesses in many places, you might consider switching to lists here.

Reply:
I just pushed a new version of only 50 lines using helper functions.
Please check it out :)